File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution
{
public:
  vector<int> findAnagrams(string s, string p)
  {
    vector<int> sHash(26, 0);
    vector<int> pHash(26, 0);
    vector<int> res;

    if (p.size() > s.size())
      return res;

    int left = 0, right = 0;
    while (right < p.size())
    {
      pHash[p[right] - 'a']++;
      sHash[s[right] - 'a']++;
      right++;
    }
    right--;
    while (right < s.size())
    {
      if (pHash == sHash)
        res.push_back(left);
      right++;
      if (right != s.size())
      {
        sHash[s[right] - 'a']++;
        sHash[s[left] - 'a']--;
      }
      left++;
    }
    return res;
  }
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function (s, p) {
  const pHash = new Array(26).fill(0);
  const sHash = new Array(26).fill(0);

  const res = [];

  if (p.length > s.length) {
    return res;
  }

  let left = 0;
  let right = 0;
  while (right < p.length) {
    pHash[p.charCodeAt(right) - 97]++;
    sHash[s.charCodeAt(right) - 97]++;
    right++;
  }
  right--;

  while (right < s.length) {
    if (pHash.join("") === sHash.join("")) {
      res.push(left);
    }
    right++;
    if (right !== s.length) {
      sHash[s.charCodeAt(right) - 97]++;
      sHash[s.charCodeAt(left) - 97]--;
    }
    left++;
  }
  return res;
};